### PR TITLE
Add customer access control middlewares

### LINKED ISSUE #119: Ensure customer is making the request for their own commits when they make POST & DELETE requests on commits
Do that by checking whether the authenticated customer is the same as the customer of the new/existing commit.
An effort to deal with non-happy paths.

### PR DESCRIPTION
**Pull request by hzxcaryn**
Closes #119

# Background
Previously, we only check for authentication of customers, and did not have access control restrictions. This means that as long as I have a valid token, I can make requests to modify another customer's commits and customer data. This is not desirable and this PR fixes this by adding a layer of access control to restricted customer endpoints. 

# Changes
- Added access control middleware that checks if customer is authorized to access/modify the resource they have requested
  * This middleware requires a prior middleware to add an authorized gpay id to `req.authorizedCustomerGPayId` before it can be checked in the `customerAccessControl` middleware. 
(Initially, I wanted `customerAccessControl` middleware to be a higher order function (as seen in the very first commit of this PR), but it was too much of a pain to mock it, and in view of the tight deadline, I decided to change it to this current pattern instead)
- Hoisted validation code in the `api` layer to _above_ the `customerAccessControl` middleware, because we don't want the part that sets `req.authorizedCustomerGPayId` to skip validations
- Added tests & a test mock for the middleware

### CODE REVIEW

File: server/src/middleware/customer-access-control.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    if (gpayIdOfCurrCustomer === undefined) {
+      throw new UnauthorizedError('User cannot be identified.');
+    }
+
+    const gpayIdOfAuthorizedCustomer = req.authorizedCustomerGPayId;
+    if (gpayIdOfAuthorizedCustomer === undefined) {
+      throw new InternalServerError('req.authorizedCustomerGPayId is not set.');
+    }
+
+    if (gpayIdOfCurrCustomer !== gpayIdOfAuthorizedCustomer) {
```
Comment (ifhuang) [inline server/src/middleware/customer-access-control.ts:56]: so gpayIdOfCurrCustomer is from spot identity api. where is gpayIdOfAuthorizedCustomer from?

File: server/src/api/customers.ts [hunk 2]
```diff
@@ -58,7 +59,17 @@ customerRouter.post(
   '/',
   customerAuth,
   async (req: Request, res: Response, next: NextFunction) => {
-    const customerData: CustomerPayload = req.body;
+    // TODO: Ensure that req.body is of type CustomerPayload
+    req.validated = req;
+    next();
+  },
+  async (req: Request, res: Response, next: NextFunction) => {
+    req.authorizedCustomerGPayId = req.validated.body.gpayId;
```
Comment (ifhuang) [inline server/src/api/customers.ts:67]: looks like req.validated.body.gpayId is also from spot api identity?
if so, then customerAccessControl is not needed here since inside it's comparing two ids from the same source?

File: server/tests/integration/customers.test.ts [hunk 3]
```diff
@@ -89,9 +99,33 @@ describe('Customers endpoints', () => {
...
+
+      const expectedCustomerData = customerFixtures.responseData?.[0];
+      const gpayId = expectedCustomerData.gpayId;
+
+      const res = await request(app).post('/customers').send({gpayId});
+
+      expect(res.status).toBe(403);
+      expect(res.body).toHaveProperty('error');
+      expect(res.body.error.message).toBe(
+        'Not allowed to modify or access requested resource.'
```
Comment (ifhuang) [inline server/tests/integration/customers.test.ts:121]: where did you set gpayid from spot identity api/token?

File: server/src/middleware/customer-access-control.ts (hunk 1 above)
Comment (hzxcaryn) [reply server/src/middleware/customer-access-control.ts:56]: @ifhuang 
`gpayIdOfCurrCustomer` is derived from the **decoded Authorization token** that is placed in Header of requests that requires customer auth
`gpayIdOfAuthorizedCustomer` is derived from the request. So for example you want to edit customer 123's data in the request. `gpayIdOfAuthorizedCustomer` will be the gpayId of customer 123.
We will then check if gpayId of customer 123 matches with the gpay id in the authorization token provided by the customer.
Basically this prevents 1 authorized customer from accessing/modifying another authorized customer's data

File: server/src/api/customers.ts (hunk 2 above)
Comment (hzxcaryn) [reply server/src/api/customers.ts:67]: @ifhuang `req.validated.body.gpayId` is the `gpayId` field in the request body. From the client Group Buy app, it is indeed retrieved from api identity, but on the server side, we cannot just trust the input from the request, because another user can just replay the request with their own token, but with my gpayId.

File: server/tests/integration/customers.test.ts (hunk 3 above)
Comment (hzxcaryn) [reply server/tests/integration/customers.test.ts:121]: @ifhuang Oh you mean for the test? It is in the customer-auth mock which is in PR #212. Basically I just added a dummy sub for the mock.
image

File: server/src/middleware/customer-access-control.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  UnauthorizedError,
+  ForbiddenError,
+  InternalServerError,
+} from '../utils/http-errors';
+
+/**
+ * Authorizes a customer.
+ * Requires req.authorizedCustomerGpayId to be set in order for this
+ * access control middleware to work.
+ * @params getGpayIdOfAuthorizedCustomer Getter function to get gpayId of
```
Comment (chshda) [inline server/src/middleware/customer-access-control.ts:29]: Where is this getGpayIdOfAuthorizedCustomer?

Comment (hzxcaryn) [reply server/src/middleware/customer-access-control.ts:29]: @chshda Ah good catch! I removed it because mocks were not cooperating for HOF middlwares, but left out this line when updating!
